FilePersistence/src/version_control/merge/ChangeGraph.cpp

```diff
@@ -762,9 +762,58 @@ bool ChangeGraph::removeDependenciesInsideNonConflictingAtomicChangeGroups()
 		if (okToRemoveDependenciesInCycle)
 		{
 			// This is a cycle which matches all conditions. Remove the dependencies between the elements.
+			// NOTE: Here we still keep dependencies such as
+			// -In case of Insertion/MoveIn, we need to make sure that it's parent exist
+			// -In case of Deletion, we need to make sure that its subtree is deleted
+			// We allow label conflicts in this case, since they will be resolved once all changes are applied
 			removedSomeDependencies = true;
 			for (auto changeToMakeIndependent : changesInCycle)
-				removeAllDependencies(changeToMakeIndependent);
+			{
+				// Nothing will depend on Deletion ( Except Deletion of parentNode )
+				// Deletion will depend on Deletion, MoveOut of children
+				// MoveIn, Insertion of children will depend on Insertion
+				// Insertion will depend on nothing ( Except Insertion of parentNode )
+				if (changeToMakeIndependent->type() != ChangeType::Deletion)
+				{
+					// Remove all forward dependencies if change is not deletion
+					bool allDependenciesRemoved = true;
+					auto thisDependsOn = dependencies_.values(changeToMakeIndependent);
+					for (auto ourDependency : thisDependsOn)
+					{
+						// If change depends on Insertion(ourDependency), then it must be Insert/MoveIn change
+						// in the subtree of element to be Inserted(by ourDependency), so dependency cannot be removed
+						// New Parent must exist for the change to happen
+						if (ourDependency->type() != ChangeType::Insertion)
+						{
+							reverseDependencies_.remove(ourDependency, changeToMakeIndependent);
+							dependencies_.remove(changeToMakeIndependent, ourDependency);
+						}
+						else
+							allDependenciesRemoved = false;
+					}
+					if (allDependenciesRemoved)	dependencies_.remove(changeToMakeIndependent);
```

Comment:
now that you're removing the change inside the `if` above, you no longer need to remove it here. Get rid of this line and of the `allDependenciesRemoved` variable.